### PR TITLE
Add type annotations

### PR DESCRIPTION
**Pull request by mdellweg**
Started to add type annotations.

This already revealed small inconsistencies (mostly around str vs bytes) and one actual bug (the `super` call).

### CODE REVIEW

**Comment by daviddavis** [comment]
I was just thinking about doing this.

**Comment by mdellweg** [comment]
It's incompatible with python2. Is that a problem?

Well, it could be done with py2, but it's a mess.

File: pulpcore/cli/common.py [hunk 1]
```diff
@@ -7,19 +9,27 @@
 
 
 class PulpJSONEncoder(json.JSONEncoder):
-    def default(self, obj):
+    def default(self, obj: Any) -> str:
         if isinstance(obj, datetime.datetime):
             return obj.isoformat()
         else:
-            super(json.JSONEncoder).default(obj)
+            return super().default(obj)
```
Comment (daviddavis) [inline pulpcore/cli/common.py:16]: Nice that you were able to catch this using typing hints.

**Comment by daviddavis** [comment]
> It's incompatible with python2. Is that a problem?

That's a good question. I don't think it is but let me confirm.

**Comment by mdellweg** [comment]
now with `--strict`. :grinning:

**Comment by mdellweg** [comment]
> > It's incompatible with python2. Is that a problem?
> 
> That's a good question. I don't think it is but let me confirm.

FWIW We don't support py2 anyway to this point:

https://github.com/pulp/pulp-cli/blob/c7ca274c947015064d295d6fc2dc299c055c4b93/.github/workflows/main.yml#L23

File: README.md [hunk 2]
```diff
@@ -15,7 +15,7 @@ options include:
 ## Configuration
 
 The CLI can be configured by using a toml file.
-By default the location of this file is `~/.config/pulp/settings.toml`. 
+By default the location of this file is `~/.config/pulp/settings.toml`.
```
Comment (daviddavis) [inline README.md:18]: :+1:

File: pulpcore/cli/core/artifact.py [hunk 3]
```diff
@@ -1,29 +1,32 @@
...
     result = ctx.obj.call("artifacts_list", parameters=params)
     ctx.obj.output_result(result)
 
 
 @artifact.command()
 @click.option("--file", type=click.File("rb"), required=True)
-@click.option("--chunk-size", default=1000000, help="Chunk size in bytes (default is 1 MB)")
+@click.option(
+    "--chunk-size", default=1000000, type=int, help="Chunk size in bytes (default is 1 MB)"
+)
```
Comment (daviddavis) [inline pulpcore/cli/core/artifact.py:27]: Why the change here?

File: pulpcore/cli/core_cli.py [hunk 4]
```diff
@@ -1,16 +1,17 @@
...
 
 @main.command()
 @click.pass_context
-def status(ctx):
-    result = ctx.obj.call("status_read")
-    ctx.obj.output_result(result)
+def status(ctx: click.Context) -> None:
+    pulp_ctx: PulpContext = ctx.find_object(PulpContext)
+    result = pulp_ctx.call("status_read")
+    pulp_ctx.output_result(result)
```
Comment (daviddavis) [inline pulpcore/cli/core_cli.py:14]: I'm curious why this changed as compared to say `orphans_delete()`?

File: pulpcore/cli/common.py [hunk 5]
```diff
@@ -42,13 +55,11 @@ def wait_for_task(self, task_href, timeout=120):
...
             if task["state"] == "canceled":
                 raise click.ClickException("Task canceled")
-            time.sleep(1)
-            timeout -= 1
             click.echo(".", nl=False, err=True)
         raise click.ClickException("Task timed out")
 
 
-def _config_callback(ctx, param, value):
+def _config_callback(ctx: click.Context, param: Any, value: str) -> None:
```
Comment (daviddavis) [inline pulpcore/cli/common.py:62]: I think param could maybe be a str but I guess it doesn't make a huge difference since we're not using it.

File: pulpcore/cli/core/artifact.py (hunk 3 above)
Comment (mdellweg) [reply pulpcore/cli/core/artifact.py:27]: It is adding `type=int` and i hope that will add input validation on the click side.
Not sure if the decorator will add type annotations to the variable passed to the command, but that would be super cool.

File: pulpcore/cli/core_cli.py (hunk 4 above)
Comment (mdellweg) [reply pulpcore/cli/core_cli.py:14]: I didn't know about `find_object` but it looks like you can embed more than one object in the context that way.
I think, this is more explicit and forgot about `orphan_delete` it seems.

File: pulpcore/cli/common.py (hunk 5 above)
Comment (mdellweg) [reply pulpcore/cli/common.py:62]: I think, `str` didn't work...

**Review by daviddavis — APPROVED** [review]
